### ISSUE TITLE
Space-control aggregation returns numeric zero for fully unobserved frames

### ISSUE THREAD

**Issue by Litju** [floodlight-sports/floodlight#187]
## Summary

When both teams have no observed player coordinates for a retained frame,
`DiscreteVoronoiModel` preserves an all-`NaN` internal ownership grid, but its
public aggregation methods return finite numeric values:

- every player: `0.0`
- Home: `0.0`
- Away: `0.0`

This converts:

```text
no territorial-control observation exists
```

into:

```text
an observed territorial-control value of zero
```

I would like maintainer guidance on treating this state as missing before
preparing a production PR.

The proposed initial repair is intentionally narrow: preserve temporal alignment
and return `NaN` for player and team control only when the internal ownership
frame contains no assigned owner.

During reproduction, I also found broader observation-validity questions around
one-team dropout, partial-player dropout, and motion-model derivative support.
Those findings are documented below so that the narrow scope is an explicit
design decision rather than an omission.

---

## Minimal reproduction

```python
import numpy as np

from floodlight import Pitch, XY
from floodlight.models.space import DiscreteVoronoiModel

pitch = Pitch.from_template("dfl", length=105, width=68, sport="football")
missing_home = XY(np.full((3, 2), np.nan), framerate=10)
missing_away = XY(np.full((3, 2), np.nan), framerate=10)

model = DiscreteVoronoiModel(
    pitch,
    mesh="square",
    xpoints=10,
    motion_model="euclidean",
)
model.fit(missing_home, missing_away)

print(np.all(np.isnan(model._cell_controls_)))
# True

print(model.player_controls()[0].property)
# currently [[0.], [0.], [0.]]

print(model.team_controls()[0].property)
# currently [[0.], [0.], [0.]]
```

The same public `0.0` behavior was reproduced for:

- Euclidean ownership
- Taki–Hasegawa ownership
- Fujimura–Sugihara ownership
- square meshes
- hexagonal meshes

---

## Current mathematical behavior

Let `G` be all mesh cells, `o_t(g)` the owner stored for cell `g` at frame `t`,
`N_H` the number of Home players, and `i` a stacked player identifier.

The current player aggregation is:

```text
C_i(t) = round(100 * count(o_t(g) == i) / |G|, 2)
```

The current team aggregation is:

```text
C_H(t) = round(100 * count(o_t(g) < N_H) / |G|, 2)
C_A(t) = round(100 * count(o_t(g) >= N_H) / |G|, 2)
```

The denominator is always the total number of grid cells.

For a completely unobserved frame, the producer retains an all-`NaN` ownership
grid. Comparisons involving `NaN` are false, so every numerator is zero while
the denominator remains nonzero:

```text
100 * 0 / |G| = 0
```

The resulting `0.0` is therefore produced by aggregation mechanics rather than
by an observed territorial partition.

---

## Shared architectural impact

All supported configurations converge on the same internal ownership array and
the same public aggregation boundary:

```text
Home/Away XY inputs
    ↓
Euclidean, Taki–Hasegawa, or Fujimura–Sugihara costs
    ↓
square or hexagonal mesh ownership
    ↓
_cell_controls_
    ↓
player_controls()
team_controls()
```

The ownership producer already preserves the fully undefined state correctly as
an all-`NaN` frame. Semantic information is lost only when `player_controls()`
and `team_controls()` convert that frame into percentages.

This means the defect affects every model/mesh path but has one shared,
surgical repair boundary.

---

## Why this matters

A finite zero normally communicates that a quantity was observed and measured
as zero. That is not the state represented here: no ownership observation
exists.

The distinction matters for temporal averages, tactical summaries, frame-wise
comparisons, downstream statistics, missing-data filtering, plotting, and
reporting.

The frame should remain present so temporal alignment is preserved. Its
numerical output should communicate that the estimate is undefined.

This is consistent with Floodlight's existing use of `NaN` in property,
geometry, resampling, plotting, and aggregation workflows.

---

## Controlled reproduction and dropout-sensitivity experiment

A deterministic synthetic experiment compared dropout-perturbed outputs with
complete-data outputs from the same trajectories and model configuration.

The experiment included:

- 78 complete-data references
- 2,220 deterministic dropout cases
- 28,032 frame-level results
- 0 failed cases
- 0 nonfatal warnings
- all three ownership models
- both mesh types
- representative mesh resolutions at `xpoints=10`, `20`, and `40`
- six static formation families
- three dynamic trajectory families
- exact clean-process reproducibility

These are implementation-specific synthetic sensitivity results. They do not
establish real-world model accuracy, provider-wide dropout behavior, or a
universal acceptable-error threshold.

<details>
<summary><strong>Principal observation-validity findings</strong></summary>

### Both teams fully unobserved

Every tested path produced:

```text
finite owner cells = 0
internal ownership = all NaN
public player outputs = 0.0
public team outputs = 0.0 / 0.0
```

This state is categorically non-estimable.

### One complete team unobserved

Current output becomes a conserved `100/0` or `0/100` partition.

Against matched complete-data references, maximum team-control error ranged
from `36.25` to `63.46` percentage points.

A numerically conserved 100% partition therefore does not establish that the
observation is scientifically complete.

### Partial-player dropout

When at least one player remains observed for each team, Floodlight partitions
the pitch among the remaining players. The resulting estimand is territorial
control conditional on the observed players.

Across tested partial-dropout cases with both teams still represented:

```text
maximum team error range: 0.00 to 60.38 percentage points
leave-one-player-out maximum: 8.46 percentage points
matched random-subset maximum: 22.19 percentage points
```

At identical 87.5% player coverage, different missing-player identities
produced `0.00` versus `6.92` percentage-point team error. Player count or
observed fraction is descriptive but cannot determine error by itself.

### Owned-cell coverage

Whenever any valid player remained, the current producer assigned every grid
cell, so `owned_cell_fraction = 1.0`. Only a completely unobserved frame
produced `owned_cell_fraction = 0.0`.

Owned-cell fraction therefore measures whether the algorithm produced any
partition, not whether the tracking observation was complete.

### Motion-model temporal support

Euclidean ownership remained frame-local.

Taki–Hasegawa and Fujimura–Sugihara use velocity preprocessing based on
neighboring frames. Missing intervals affected one otherwise position-observed
frame immediately before and one immediately after the interval in the tested
implementation.

Position presence and derivative support are therefore separate validity
dimensions.

</details>

---

## Observation states uncovered by the investigation

| Observation state | Current public behavior | Interpretation |
|---|---|---|
| Fully observed | Current percentages | Complete-data model estimate |
| Partial Home dropout | Conserved conditional repartition | Conditional on observed Home players |
| Partial Away dropout | Conserved conditional repartition | Conditional on observed Away players |
| Partial bilateral dropout | Conserved conditional repartition | Conditional on both observed subsets |
| Home entirely unobserved | `0/100` | Away-only partition |
| Away entirely unobserved | `100/0` | Home-only partition |
| Both teams entirely unobserved | `0/0` | No estimable ownership observation |
| Motion-neighbor affected | Finite percentages | Positions present, derivative support affected |

The proposed initial repair addresses only the categorically unambiguous
aggregation defect:

```text
both teams fully unobserved
→ no assigned owner exists
→ return NaN
```

The other states require broader policy or API design.

---

## Recommended initial repair: no-owner `NaN` guard

A disposable local prototype used one private shared predicate derived directly
from the existing ownership grid:

```python
def _frames_have_assigned_owners(self) -> np.ndarray:
    return np.any(np.isfinite(self._cell_controls_), axis=(1, 2))
```

Both `player_controls()` and `team_controls()` use the same predicate after
their existing percentage calculations.

The predicate name intentionally states only that an owner exists. It does not
claim that the frame is fully observed, scientifically complete, or valid for
every analytical purpose.

### Proposed semantics

| Internal ownership state | Proposed public behavior |
|---|---|
| At least one assigned owner | Preserve all current percentages exactly |
| No assigned owner | Every retained player and both teams become `NaN` |
| Frame indexing and count | Preserved |
| Denominator and rounding | Unchanged |
| Ownership producer | Unchanged |
| Motion preprocessing | Unchanged |
| Grid construction | Unchanged |
| Public signatures and return types | Unchanged |

---

## Local prototype feasibility evidence

The disposable prototype modified only `floodlight/models/space.py`.

Production diff:

```text
12 additions
0 removals
1 new private symbol
0 new public symbols
```

Results:

- 174/174 assigned-frame outputs were exactly identical
- 0 assigned frames changed
- all 18 fully missing model/mesh results changed from zero to `NaN`
- all fully missing frames remained retained and aligned
- all 90 one-team and partial-dropout comparisons remained exactly unchanged
- all 18 empty-team boundary combinations gained no new exception or shape regression
- focused tests: 66/66 passed
- existing space-control tests: 18/18 passed
- all model tests: 134/134 passed
- downstream temporal/zone tests: 28/28 passed
- complete repository suite: 345/345 passed
- Black, Flake8, and `git diff --check` passed

A controlled mutation replaced the four `NaN` assignments with `0.0`. All 12
selected missing-output regression tests failed on the intended assertions.
Restoring `NaN` caused all 12 to pass again.

The prototype was not committed, pushed, or proposed for merge.

---

## Solution directions considered

These directions address different contracts. They are not equivalent
implementations of the same repair.

| Direction | Problem addressed | Additional state or API | Maintenance surface | Proposed status |
|---|---|---|---|---|
| No-owner `NaN` guard | Fully undefined ownership becomes numeric zero | Existing ownership grid only | Low | Recommended initial PR |
| One-team observation guard | Missing opposition becomes `100/0` | Retained input-observation state | Medium | Separate policy discussion |
| Observation-quality sidecar | Missingness is not publicly communicated | New metadata/public lifecycle | High initially | Separate RFC/design issue |
| Finite-owner denominator | Attempts to exclude undefined cells | Formula change | Adds little value | Rejected |
| Motion-support validity | Derivatives depend on missing neighboring frames | Velocity-stencil support state | Separate contract | Follow-up design |

### Recommended: no-owner `NaN` guard

Advantages:

- repairs the only categorically undefined public output;
- uses the missing state already preserved internally;
- requires no new public interface or retained input state;
- leaves every assigned frame unchanged;
- applies consistently to all models and meshes;
- has a small, testable, shared implementation boundary.

Limitations are deliberate: one-team dropout remains `100/0`, partial-player
dropout remains conditional, and motion derivative support remains undisclosed.

### Broader policy alternative: one-team observation guard

A stronger policy could require at least one observed player from each team.
Potentially, if either team is entirely unobserved, all player and team outputs
for that frame would become `NaN`.

Scientific advantage:

- prevents missing-opposition `100/0` from appearing as a complete tactical
  estimate.

Maintenance and compatibility cost:

- percentages alone cannot distinguish legitimate complete `100/0` from
  missing-team `100/0`;
- input-observation state must be retained or propagated from `fit()`;
- deliberately one-sided inputs and tracking dropout need distinct semantics;
- existing one-team behavior changes;
- broader player/team consistency rules are required.

The experiment strongly motivates this discussion, but the policy belongs to
maintainers and should not be combined silently with the narrow repair.

### Longer-term API alternative: observation-quality sidecar

Potential information could include:

```text
position_observation_status
observed_home_players
observed_away_players
home_observed_fraction
away_observed_fraction
both_teams_observed
motion_support_valid
```

Advantages:

- communicates incomplete observation without automatically discarding every
  conditional estimate;
- distinguishes per-team observation completeness;
- can separately disclose position presence and derivative support.

Design and maintenance cost:

- requires a new public contract;
- may affect slicing, copying, resampling, pandas conversion, serialization,
  plotting, and downstream consumers;
- counts and fractions are identity-blind and insufficient as validity criteria;
- overall coverage hides team asymmetry;
- owned-cell fraction is degenerate;
- one flat enum may be insufficient.

At least three independent dimensions may coexist:

```text
position-observation completeness
temporal missingness pattern
motion-derivative support
```

This is better treated as a separate RFC or design issue.

### Rejected: finite-owned-cell denominator

Replacing `owned count / all cells` with `owned count / finite owner cells` was
rejected because the current producer normally assigns every grid cell whenever
any valid player remains.

Consequently, it does not detect one-team or partial-player dropout, still
returns a conserved conditional 100% partition, changes the aggregation
formula, and produces missingness for the fully undefined case only indirectly
through a zero denominator.

The explicit no-owner guard communicates the intended semantics more clearly
and with less technical debt.

### Separate contract: motion-support validity

Position observation and derivative support are not the same state.

Taki–Hasegawa and Fujimura–Sugihara may depend on missing coordinates in
neighboring frames even when the current frame's positions are finite. A future
motion-support field requires a precise definition of derivative-stencil
support, potentially at player rather than frame level.

This should not be solved by changing or interpolating velocities in this
proposal.

---

<details>
<summary><strong>Adjacent findings intentionally excluded from the initial repair</strong></summary>

The investigation also reproduced or identified adjacent technical contracts:

- half-missing coordinate pairs can bypass the full-frame skip and raise an
  all-`NaN` reduction error;
- empty teams can produce a malformed empty player-property shape `(0,)`;
- exact ownership ties favor the first stacked player identifier;
- input validation is permissive for sequence lengths, framerates, coordinate
  pairs, and empty teams;
- motion-model missingness can affect neighboring valid frames.

These findings are not required to repair fully no-owner aggregation semantics
and should not be combined into the initial PR without explicit maintainer
agreement.

</details>

---

## Historical context

The all-cell aggregation behavior was introduced before Floodlight began
preserving all-`NaN` ownership frames.

A later change taught the ownership producer to skip completely missing frames
and retain their internal `NaN` state, but the public player/team aggregators
were not updated to preserve that missingness.

The current `0/0` behavior therefore appears likely incidental rather than an
explicitly documented missing-output policy.

---

## Why the proposed initial scope remains narrow

The broader findings do not weaken the no-owner repair. They explain why it
should not silently define unrelated policy.

The proposed initial PR would address only:

```text
no assigned owner exists
→ expose NaN for every player and both teams
```

It would not decide one-team-only policy, partial-player semantics, observation
metadata, motion derivative support, or malformed-input validation. Those
questions have materially different state, API, and maintenance requirements
and should become separate follow-up issues or an RFC after maintainer feedback.

---

## Explicit exclusions

The proposed initial repair would not:

- change one-team or partial-dropout behavior;
- change motion or velocity preprocessing;
- add observation-quality metadata;
- add or change public method signatures;
- change `PlayerProperty` or `TeamProperty`;
- validate half-missing coordinate pairs;
- repair empty-team output shapes;
- change exact-tie behavior;
- change percentage denominators or rounding;
- change mesh generation or model equations;
- interpolate missing positions;
- omit frames.

---

## Proposed acceptance criteria

- no-owner frames remain temporally aligned;
- every retained player and both teams are `NaN` on no-owner frames;
- every assigned-frame player and team output remains exactly unchanged;
- player and team aggregation share one private no-owner rule;
- all three models and both mesh types are covered;
- valid/missing/valid sequences preserve order and adjacent valid output;
- one-team and partial-dropout behavior is not changed;
- empty-team behavior is not worsened;
- output classes, normal shapes, names, and framerate remain unchanged;
- no public API, denominator, rounding, producer, motion, grid, or tie-policy
  change is introduced;
- focused tests and the complete suite pass;
- lint and `git diff --check` pass;
- regression tests fail if `NaN` is replaced with zero.

---

## Maintainer questions

1. Should a frame with no assigned owner expose `NaN` for every player and both
   team-control outputs instead of numeric zero?
2. Is the aggregation-only no-owner guard an acceptable scope for an initial PR?
3. Should one-team observation policy and observation-quality metadata remain
   separate follow-up discussions?
4. Is a changelog entry expected for this numerical missingness correction?
5. Are there known downstream compatibility expectations for fully unobserved
   space-control frames that the eventual PR should preserve?

**Comment by Litju** [comment]
Hi @manuba95 — I wanted to follow up on the proposed missingness contract here before preparing a production PR.

The local prototype preserves every assigned-frame output exactly and changes only frames where the internal ownership grid has no assigned owner, returning `NaN` instead of numeric `0.0` for player and team control.

Would that aggregation-only `NaN` guard be an acceptable initial scope? I would keep one-team dropout, partial-player dropout, motion support, and observation-quality metadata as separate follow-up discussions.
